File: backend/app/schemas/team.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.schemas.auth import UserPublic

TeamRoleValue = Literal["TEAM_ADMIN", "TEAM_MEMBER"]
TeamInvitationStatusValue = Literal["PENDING", "ACCEPTED", "EXPIRED", "CANCELLED"]


class TeamCreateRequest(BaseModel):
    name: str = Field(min_length=2, max_length=50)
    description: str | None = Field(default=None, max_length=500)


class TeamPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    created_by_id: int
    created_at: datetime


class TeamMemberPublic(BaseModel):
    id: int
    team_id: int
    user: UserPublic
    role: TeamRoleValue
    created_at: datetime


class TeamInvitationCreateRequest(BaseModel):
    email: EmailStr
    role: TeamRoleValue = "TEAM_MEMBER"


class TeamInvitationPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    team_id: int
    email: EmailStr
    role: TeamRoleValue
    status: TeamInvitationStatusValue
    invited_by_id: int
    accepted_by_id: int | None
    expires_at: datetime
    accepted_at: datetime | None
    cancelled_at: datetime | None
    created_at: datetime


class TeamMemberRoleUpdateRequest(BaseModel):
    role: TeamRoleValue


class ActionResponse(BaseModel):
    status: str

File: frontend/src/types/team.ts
import type { UserPublic } from "./auth";

export type TeamRole = "TEAM_ADMIN" | "TEAM_MEMBER";
export type TeamInvitationStatus = "PENDING" | "ACCEPTED" | "EXPIRED" | "CANCELLED";

export interface Team {
  id: number;
  name: string;
  description: string | null;
  created_by_id: number;
  created_at: string;
}

export interface TeamMember {
  id: number;
  team_id: number;
  user: UserPublic;
  role: TeamRole;
  created_at: string;
}

export interface TeamInvitation {
  id: number;
  team_id: number;
  email: string;
  role: TeamRole;
  status: TeamInvitationStatus;
  invited_by_id: number;
  accepted_by_id: number | null;
  expires_at: string;
  accepted_at: string | null;
  cancelled_at: string | null;
  created_at: string;
}

export interface TeamCreatePayload {
  name: string;
  description?: string | null;
}

export interface TeamInvitationCreatePayload {
  email: string;
  role: TeamRole;
}

export interface TeamMemberRoleUpdatePayload {
  role: TeamRole;
}
